include/volume/volume_cooling_gaussians_mass_matrix.hpp

```diff
@@ -0,0 +1,528 @@
+// VolEsti (volume computation and sampling library)
+
+// Copyright (c) 2012-2020 Vissarion Fisikopoulos
+// Copyright (c) 2018-2020 Apostolos Chalkis
+
+//Contributed and/or modified by Apostolos Chalkis, as part of Google Summer of Code 2018 program.
+
+// Licensed under GNU LGPL.3, see LICENCE file
+
+#ifndef VOLUME_COOLING_GAUSSIANS_HPP
+#define VOLUME_COOLING_GAUSSIANS_HPP
+
+//#define VOLESTI_DEBUG
+
+#include <iterator>
+#include <vector>
+#include <list>
+#include <math.h>
+#include <chrono>
+
+#include "cartesian_geom/cartesian_kernel.h"
+#include "random_walks/gaussian_helpers.hpp"
+#include "random_walks/gaussian_ball_walk.hpp"
+#include "random_walks/gaussian_cdhr_walk.hpp"
+#include "random_walks/gaussian_hamiltonian_monte_carlo_exact_walk.hpp"
+#include "sampling/random_point_generators.hpp"
+#include "volume/math_helpers.hpp"
+#include "Eigen/Eigen"
+
+
+/////////////////// Helpers for random walks
+
+template <typename WalkType>
+struct update_delta
+{
+    template <typename NT>
+    static void apply(WalkType, NT) {}
+};
+
+template <typename Polytope, typename RandomNumberGenerator>
+struct update_delta<GaussianHamiltonianMonteCarloExactWalk::Walk<Polytope, RandomNumberGenerator>>
+{
+    template <typename NT>
+    static void apply(GaussianHamiltonianMonteCarloExactWalk::Walk<Polytope, RandomNumberGenerator> & walk,
+                      NT delta)
+    {
+        walk.update_delta(delta);
+    }
+};
+
+template <typename Polytope, typename RandomNumberGenerator>
+struct update_delta<GaussianBallWalk::Walk<Polytope, RandomNumberGenerator>>
+{
+    template <typename NT>
+    static void apply(GaussianBallWalk::Walk<Polytope, RandomNumberGenerator> & walk,
+                      NT delta)
+    {
+        walk.update_delta(delta);
+    }
+};
+
+
+////////////////////////////// Algorithms
+
+// Gaussian Anealling
+
+// Implementation is based on algorithm from paper "A practical volume algorithm",
+// Springer-Verlag Berlin Heidelberg and The Mathematical Programming Society 2015
+// Ben Cousins, Santosh Vempala
+
+// Compute the first variance a_0 for the starting gaussian
+template <typename Polytope, typename NT>
+void get_first_gaussian(Polytope const& P,
```

Comment:
there function as well as all the following are redefined in https://github.com/GeomScale/volesti/blob/develop/include/volume/volume_cooling_gaussians.hpp 
Please try to reuse as much code as possible. This file here should provide a new function `volume_general_cooling_gaussians()` that reuses functions from https://github.com/GeomScale/volesti/blob/develop/include/volume/volume_cooling_gaussians.hpp and some new functions (with the implementation of the general mass matrix).